### ISSUE TITLE
Soundness issue in Ctype.non_aliasable

### ISSUE THREAD

**Issue by stedolan** [ocaml/ocaml#11492]
The typechecker assumes that abstract types from the current module (that is, whose paths contain no dots) must have come from an empty definition like `type t`, and thus cannot alias any other type.

This turns out not to be true: they can also come from `include F (struct ... end)`. This leads to unsoundness, since the typechecker will accept certain matches as exhaustive because it believes abstract types from the current module to be non-aliasable. Concretely, the following program segfaults, because the compiler believes that `t` and `int` cannot alias and so accepts the pattern in `f` as exhaustive:

```ocaml
type _ is_int = Int : int is_int

module F (X : sig end) : sig
  type t
  val int : t is_int
end = struct
  type t = int
  let int = Int
end

include F (struct end)

let f : (t is_int, string) result -> string =
  fun (Error s) -> s

let () =
  print_endline (f (Ok int))
```

I think the fix is just to remove the call to `in_current_module` (and carefully examine other uses of this function), and replace the call to `in_pervasives` that used to be there. This is not a backwards-compatible change, however, and I'm not sure how much code relies on the current behaviour.

**Comment by lpw25** [comment]
See #8900

**Comment by yminsky** [comment]
Would fixing #8900 resolve this soundness issue? Seems like a good argument for going back to that PR and pushing the fix through. Curious what @gasche thinks.

**Comment by garrigue** [comment]
Thanks for spotting this. The language is so rich that we end up forgetting corner cases at times.
Did you find it in the wild?
My immediate impression is that it is not so urgent to fix this, and that we can wait for an agreeable solution to #8900, but that we shouldn't delay too much.
My argument is, as before, that I agree that types for the current modules shouldn't be handled specially, but that there is no reason either to handle types from Pervasives specially, so that we need some way at least to give some types a unique identity. My [nominal types](https://github.com/ocaml/RFCs/blob/5048c15033688319759b16c7ce3dfa47816985a7/rfcs/nominal-types.md) proposal may take some time to resurrect, so we could start with something simpler that can be extended later.

**Comment by gasche** [comment]
@yminsky #8900 is not only a fix, it's a language change that will break code. There is consensus that the current behavior should be changed, and I think the present issue is only going to strengthen this consensus. In https://github.com/ocaml/ocaml/pull/8900#issuecomment-1029959041 I asked that people (in particular core maintainers) interested in this issue take the time to work out a consensus / a decision on @garrigue's 2019 proposal to replace the current behavior. I still believe that we should collectively do this. @lpw25 did good work in this direction, and @alainfrisch as well, but that does not seem enough to actually make a decision?

**Comment by yminsky** [comment]
I saw the discussion, but it doesn't really change my view, which is that the special treatment of locally defined abstract types is misfeature with no real upsides, and should be simply removed, with no other changes. Given that there was a test on opam that showed no real issues, I would have thought the right path would be just to remove it.

I hear your point about teaching materials, since I discussed the issue in the GADT chapter of Real World OCaml, and run into exactly this issue. But the current behavior is confusing in addition to being unsound, and I don't see any way of writing a tutorial that relies on the current special-casing that doesn't make matters worse. Blocking the fix on a better solution doesn't add up for me.

So all in, I'd advocate for fixing the special-casing immediately, thus fixing the soundness bug, and working on @garrigue's proposal independently.

**Comment by lpw25** [comment]
This bug has just popped up for us again. I think that we're going to go ahead and merge #8900 in our internal compiler to avoid this issue. I think that we should either do the same upstream, or if people want to keep waiting on "nominal types" then they should propose a fix to the soundness bug for the meantime.

**Comment by yallop** [comment]
I agree that we should simply go ahead and remove the special treatment of types in the current module.  Even with nominal types, we'll still need something like #8900 to fix the soundness issue reported here, so there's no benefit in waiting.  For the moment we can continue to treat pervasives specially; if we introduce nominal types, then we can eliminate that special treatment, too.

**Comment by garrigue** [comment]
If everybody thinks so, I'm fine with that.
Thinking of it, we could have done that in 5.0. Feels less good to do it later.
Probably the biggest part of the change is to fix the test suite :-)
Actual programs should not depend on it anyway.

**Comment by stedolan** [comment]
Like #8900, this is fixed in OCaml 5.5.
